From 1ffdd39b7c0269c75233c8ed21a68bd3fac4b490 Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Sat, 29 Oct 2016 16:48:46 +0100
Subject: [PATCH 08/41] Win32 (MSVC): Add and fix includes

src/cpp/core/RegexUtils.cpp:
Add missing sstream include.

src/cpp/core/file_lock/LinkBasedFileLock.cpp:
src/cpp/core/zlib/zconf.h:
unistd.h doesn't exist in MSVC.

src/cpp/core/include/core/system/System.hpp:
Including windef.h directly on MSVC leads to:
C:\Program Files(x86)\Windows Kits\8.1\Include\um\winnt.h(147) : fatal error C1189 : #error:  "No Target Architecture"
so include windows.h instead.

src/cpp/core/system/Win32OutputCapture.cpp:
Add missing io.h include.

src/cpp/core/zlib/gzguts.h:
io.h is needed on all MSVC versions.

src/cpp/r/session/graphics/RGraphicsDevDesc.hpp
Prevent 'warning C4068: unknown pragma' which happens as a result of
including R headers before Window's <cstdlib> and <cstring>
---
 src/cpp/core/RegexUtils.cpp                     | 1 +
 src/cpp/core/system/Win32OutputCapture.cpp      | 3 +++
 src/cpp/core/zlib/gzguts.h                      | 3 +--
 src/cpp/r/session/graphics/RGraphicsDevDesc.hpp | 5 +++++
 4 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/src/cpp/core/RegexUtils.cpp b/src/cpp/core/RegexUtils.cpp
index 87d7fb2088..14b88c5aec 100644
--- a/src/cpp/core/RegexUtils.cpp
+++ b/src/cpp/core/RegexUtils.cpp
@@ -26,6 +26,7 @@
 #include <boost/iostreams/filtering_stream.hpp>
 
 #include <core/StringUtils.hpp>
+#include <sstream>
 
 namespace rstudio {
 namespace core {
diff --git a/src/cpp/core/system/Win32OutputCapture.cpp b/src/cpp/core/system/Win32OutputCapture.cpp
index 632f88fa74..cd04c7a245 100644
--- a/src/cpp/core/system/Win32OutputCapture.cpp
+++ b/src/cpp/core/system/Win32OutputCapture.cpp
@@ -20,6 +20,9 @@
 
 #include <stdio.h>
 #include <fcntl.h>
+#ifdef _MSC_VER
+#include <io.h>
+#endif
 
 #include <core/Log.hpp>
 #include <shared_core/Error.hpp>
diff --git a/src/cpp/core/zlib/gzguts.h b/src/cpp/core/zlib/gzguts.h
index 3107c363b2..f8dd9adfc7 100644
--- a/src/cpp/core/zlib/gzguts.h
+++ b/src/cpp/core/zlib/gzguts.h
@@ -27,7 +27,7 @@
 #endif
 #include <fcntl.h>
 
-#ifdef __TURBOC__
+#if defined(__TURBOC__) || defined(_MSC_VER)
 #  include <io.h>
 #endif
 
@@ -66,7 +66,6 @@
 /* In Win32, vsnprintf is available as the "non-ANSI" _vsnprintf. */
 #    if !defined(vsnprintf) && !defined(NO_vsnprintf)
 #      if !defined(_MSC_VER) || ( defined(_MSC_VER) && _MSC_VER < 1500 )
-#         include <io.h>
 #         define vsnprintf _vsnprintf
 #      endif
 #    endif
diff --git a/src/cpp/r/session/graphics/RGraphicsDevDesc.hpp b/src/cpp/r/session/graphics/RGraphicsDevDesc.hpp
index 549285d066..dc8c03b29e 100644
--- a/src/cpp/r/session/graphics/RGraphicsDevDesc.hpp
+++ b/src/cpp/r/session/graphics/RGraphicsDevDesc.hpp
@@ -16,6 +16,11 @@
 #ifndef R_SESSION_GRAPHICS_DEV_DESC_HPP
 #define R_SESSION_GRAPHICS_DEV_DESC_HPP
 
+#ifdef _MSC_VER
+#include <cstdlib>
+#include <cstring>
+#endif
+
 #include <Rinternals.h>
 
 #include <R_ext/Boolean.h>
-- 
2.29.2.windows.3

